dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/EndpointErrorDecoder.java

```diff
@@ -0,0 +1,250 @@
+import com.google.common.io.CharStreams;
+import com.google.common.net.HttpHeaders;
+import com.google.common.primitives.Longs;
+import com.palantir.conjure.java.api.errors.QosException;
+import com.palantir.conjure.java.api.errors.QosReason;
+import com.palantir.conjure.java.api.errors.QosReasons;
+import com.palantir.conjure.java.api.errors.QosReasons.QosResponseDecodingAdapter;
+import com.palantir.conjure.java.api.errors.RemoteException;
+import com.palantir.conjure.java.api.errors.SerializableError;
+import com.palantir.conjure.java.api.errors.UnknownRemoteException;
+import com.palantir.conjure.java.dialogue.serde.Encoding.Deserializer;
+import com.palantir.conjure.java.serialization.ObjectMappers;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.TypeMarker;
+import com.palantir.logsafe.Arg;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.SafeLoggable;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.logsafe.exceptions.SafeExceptions;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Extracts the error from a {@link Response}.
+ * <p>If the error's name is in the {@link #errorNameToJsonDeserializerMap}, this class attempts to deserialize the
+ * {@link Response} body as JSON, to the error type. Otherwise, a {@link RemoteException} is thrown. If the
+ * {@link Response} does not adhere to the expected format, an {@link UnknownRemoteException} is thrown.
+ *
+ * @param <T> the base type of the endpoint response. It's a union of the result type and all the error types.
+ */
+final class EndpointErrorDecoder<T> {
+    private static final SafeLogger log = SafeLoggerFactory.get(EndpointErrorDecoder.class);
+    private static final ObjectMapper MAPPER = ObjectMappers.newClientObjectMapper();
+    private final Map<String, Encoding.Deserializer<? extends T>> errorNameToJsonDeserializerMap;
+
+    EndpointErrorDecoder(
+            Map<String, TypeMarker<? extends T>> errorNameToTypeMap, Optional<Encoding> maybeJsonEncoding) {
+        this.errorNameToJsonDeserializerMap = maybeJsonEncoding
+                .<Map<String, Encoding.Deserializer<? extends T>>>map(
+                        jsonEncoding -> errorNameToTypeMap.entrySet().stream()
+                                .collect(Collectors.toMap(
+                                        Map.Entry::getKey, entry -> jsonEncoding.deserializer(entry.getValue()))))
+                .orElseGet(Collections::emptyMap);
```

Comment:
I might simplify this to avoid streams, for readability using `ImmutableMap.copyOf(Maps.transformValues(map, jsonEncoding::deserializer))` (immutable map copy is required because `transformValues` does a lazy transformation)